### PR TITLE
[SPARK-16313][SQL][BRANCH-1.6] Spark should not silently drop exceptions in file listing

### PR DESCRIPTION
**Pull request by yhuai**
## What changes were proposed in this pull request?

Spark silently drops exceptions during file listing. This is a very bad behavior because it can mask legitimate errors and the resulting plan will silently have 0 rows. This patch changes it to not silently drop the errors.

After making partition discovery not silently drop exceptions, HiveMetastoreCatalog can trigger partition discovery on empty tables, which cause FileNotFoundExceptions (these Exceptions were dropped by partition discovery silently). To address this issue, this PR introduces two **hacks** to workaround the issues. These two hacks try to avoid of triggering partition discovery on empty tables in HiveMetastoreCatalog.
## How was this patch tested?

Manually tested.

**Note: This is a backport of https://github.com/apache/spark/pull/13987**

### CODE REVIEW

**Comment by rxin** [comment]
Wouldn't this create a ton of warnings during table creation?

**Comment by yhuai** [comment]
I think there will be one warning when we create a table. Or maybe there is no warning during table creation because the refresh is called lazily.

**Comment by yhuai** [comment]
test this please

**Comment by yhuai** [comment]
Let me see if we can have a flag to determine if we want to swallow the FNF (like what https://github.com/apache/spark/pull/13987/files does).

**Comment by yhuai** [comment]
test this please

**Comment by yhuai** [comment]
test this please

**Comment by yhuai** [comment]
test this please

**Comment by yhuai** [comment]
tes this please

**Comment by yhuai** [comment]
cc @marmbrus

**Comment by marmbrus** [comment]
LGTM

**Comment by yhuai** [comment]
let me take another look to see if there is a better change.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 1]
```diff
@@ -273,6 +273,22 @@ private[hive] class HiveMetastoreCatalog(val client: ClientInterface, hive: Hive
...
+          // an empty table, the dir of the table has not been created, which will
+          // cause a FileNotFoundException. So, at here we will catch the FileNotFoundException
+          // and return false.
+          hadoopFsRelation.partitionColumns.nonEmpty
+        } catch {
+          case _: java.io.FileNotFoundException =>
+            false
+        }
+      case _ => false
+    }
```
Comment (yhuai) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: This function is equivalent with `val resolvedRelation = dataSource.resolveRelation(checkPathExist = false)` in 2.0 (https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala#L427).

**Comment by yhuai** [comment]
@rxin I think this version is the minimal change. Since the partition discovery logic in inside HadoopFsRelation in 1.6 and the refresh is triggered by using lazy val, passing a flag down will introduce lots of changes.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala [hunk 2]
```diff
@@ -273,6 +273,20 @@ private[hive] class HiveMetastoreCatalog(val client: ClientInterface, hive: Hive
         serdeProperties = options)
     }
 
+    def hasPartitionColumns(relation: HadoopFsRelation): Boolean = {
+      try {
+        // Calling hadoopFsRelation.partitionColumns will trigger the refresh call of
```
Comment (rxin) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: I'd add to the comment that this is a hack for [SPARK-16313][SQL][BRANCH-1.6] Spark should not silently drop exceptions in file listing

**Comment by rxin** [comment]
LGTM other than that.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala (hunk 2 above)
Comment (yhuai) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala]: Done

**Comment by yhuai** [comment]
Thank you! I am merging this PR to branch 1.6.

**Pull request closed without merging**